models/exposure.py

```diff
@@ -1,7 +1,238 @@
 
-from models.base import Base
+import inspect
+from collections import defaultdict
 
+import numpy as np
+
+import sqlalchemy as sa
+from sqlalchemy import orm
+from sqlalchemy.dialects.postgresql import JSONB
+
+from astropy.coordinates import SkyCoord
+
+from models.base import Base, SpatiallyIndexed
+import models.instruments
+
+from pipeline.utils import normalize_header_key
+
+class CCD_Data:
+    """
+    A helper class that lazy loads the CCD data from the database.
+    When requesting one of the CCD IDs it will fetch that data from
+    disk and store it in memory.
+    To clear the memory cache, call the clear_cache() method.
+    """
+    def __init__(self, filename, inst):
+        self.filename = filename
+        self.instrument_object = inst
+        self._data = defaultdict(lambda: None)
+
+    def __getitem__(self, ccd_id):
+        if self._data[ccd_id] is None:
+            self._data[ccd_id] = self.instrument_object.load(self.filename, ccd_id)
+        return self._data[ccd_id]
+
+    def __setitem__(self, ccd_id, value):
+        self._data[ccd_id] = value
+
+    def clear_cache(self):
+        self._data = defaultdict(lambda: None)
+
+
+class Exposure(Base, SpatiallyIndexed):
 
-class Exposure(Base):
     __tablename__ = "exposures"
 
+    header = sa.Column(JSONB, nullable=False, default={}, doc="Header of the raw exposure. ")
+
+    filename = sa.Column(sa.Text, nullable=False, index=True, unique=True, doc="Filename for raw exposure. ")
+
+    mjd = sa.Column(
+        sa.Double,
+        nullable=False,
+        index=True,
+        doc="Modified Julian date of the exposure. (MJD=JD-2400000.5)"
+    )
+
+    exp_time = sa.Column(sa.Float, nullable=False, index=True, doc="Exposure time in seconds")
+
+    filter = sa.Column(sa.Text, nullable=False, index=True, doc="Filter name")
+
+    instrument = sa.Column(sa.Text, nullable=False, index=True, doc="Instrument name")
+
+    telescope = sa.Column(sa.Text, nullable=False, index=True, doc="Telescope name")
+
+    num_ccds = sa.Column(sa.Integer, nullable=False, index=False, doc="Number of CCDs / sections in the full field. ")
+
+    width = sa.Column(sa.Integer, nullable=False, index=False, doc="Width of each CCD in pixels. ")
+
+    height = sa.Column(sa.Integer, nullable=False, index=False, doc="Height of each CCD in pixels. ")
+
+    project = sa.Column(sa.Text, index=True, nullable=False, doc='Name of the project, (could also be a proposal ID). ')
+
+    target = sa.Column(sa.Text, index=True, nullable=False, doc='Name of the target object or field id. ')
+
+    gallat = sa.Column(sa.Double, index=True, doc="Galactic latitude of the target. ")
+
+    gallon = sa.Column(sa.Double, index=False, doc="Galactic longitude of the target. ")
+
+    ecllat = sa.Column(sa.Double, index=True, doc="Ecliptic latitude of the target. ")
+
+    ecllon = sa.Column(sa.Double, index=False, doc="Ecliptic longitude of the target. ")
+
+    def __init__(self, *args, **kwargs):
```

Comment:
I don't think I've used an explicit __init__ with an SQLAlchemy class before.  I'm a little confused here, though.

I would expect this __init__ to be called every time an object is instantiated.  If so, then it doesn't make sense to have lines like `self.num_ccds = self.get_instrument_object().get_num_ccds()`, as it's replacing a field that was read from the database with something calcualted.  If __init__ was only called when the database record was first created, then this would make sense, but that's at odds with general python __init__.

Reply:
OK, I read up a bit, and indeed __init__ is not called when the object is pulled out of the database.  Which is very weird to me, I have to admit.